autobahn/websocket/interfaces.py

```diff
@@ -417,6 +417,22 @@ def onConnect(self, request_or_response):
            You can also return a Deferred/Future that resolves/rejects to the above.
         """
 
+    @public
+    @abc.abstractmethod
+    def onConnecting(self, transport_details):
+        """
+        This method is called when we've connected, but before the handshake is done.
+
+        :param transport_details: information about the transport.
+        :type transport_details: :class:`autobahn.websocket.types.TransportDetails`
+
+        :returns: A
+            :class:`autobahn.websocket.types.ConnectingRequest`
+            instance is returned to indicate which options should be
+            used for this connection. If you wish to use the default
+            behavior, `None` may be returned (this is the default).
+        """
+
```

Comment:
Are all non-subclassing `IWebSocketChannel` implementations broken now, because they lack this method?  Or is the only supported use through subclassing the ABC?

Reply:
yeah, people that implemented their own sub-classing stuff have broken classes now until they add the new callback (which is strictly expected by the calling library code). not sure if anyone has done that (wrote own classes for channel). adding dynamic attribute checking to work around that .. not sure if it is worth.

Reply:
Well, but also the only way to "implement" an ABC is by subclassing it, isn't it? So then they'll get this empty method (which returns `None` and thus "asks for defaults"). I should double-check that, but I'm fairly certain that's how ABCs work...

Reply:
Hrm, I think I'm wrong here ... "A class containing at least one method declared with this decorator that hasn't been overridden yet cannot be instantiated. Such methods may be called from the overriding method in the subclass (using super or direct invocation)."

That's from PEP3119. So, I *think* what that means is that yes, anyone who created their own classes from scratch and declared the `IWebSocketChannel` "interface" via subclassing will now be broken (because they haven't implemented `onConnecting`). A quick test in the repl confirms this -- you'll get `TypeError: Can't instantiate abstract class Foo with abstract methods onClose, onConnect, onConnecting, onMessage, onOpen, onPing, onPong, sendClose, sendMessage, sendPing, sendPong`.